client/connectdlg_common.cpp

```diff
@@ -285,299 +233,67 @@ bool client_start_server(void)
-                                  sizeof(scenscmdline));
-
   if (are_deprecation_warnings_enabled()) {
-    depr = " --warnings";
-  } else {
-    depr = "";
+    arguments << "--warnings";
   }
   if (ruleset != NULL) {
-    fc_snprintf(rsparam, sizeof(rsparam), " --ruleset %s", ruleset);
-  } else {
-    rsparam[0] = '\0';
+    arguments << "--ruleset" << ruleset;
   }
 
-  fc_snprintf(options, sizeof(options),
-              "-p %d --bind localhost -q 1 -e%s%s%s --saves \"%s\" "
-              "--scenarios \"%s\" -A none %s%s",
-              internal_server_port, logcmdline, scriptcmdline,
-              savefilecmdline, savescmdline, scenscmdline, rsparam, depr);
-#ifdef FREECIV_DEBUG
-#ifdef FREECIV_WEB
-  fc_snprintf(cmdline1, sizeof(cmdline1), "./server/freeciv-web %s",
-              options);
-#else  /* FREECIV_WEB */
-  fc_snprintf(cmdline1, sizeof(cmdline1), "./fcser %s", options);
-  fc_snprintf(cmdline2, sizeof(cmdline2), "./server/freeciv-server %s",
-              options);
-#endif /* FREECIV_WEB */
-#endif /* FREECIV_DEBUG */
-#ifdef FREECIV_WEB
-  fc_snprintf(cmdline3, sizeof(cmdline2), BINDIR "/freeciv-web %s", options);
-  fc_snprintf(cmdline4, sizeof(cmdline3), "freeciv-web %s", options);
-#else  /* FREECIV_WEB */
-  fc_snprintf(cmdline3, sizeof(cmdline3), BINDIR "/freeciv-server %s",
-              options);
-  fc_snprintf(cmdline4, sizeof(cmdline4), "freeciv-server %s", options);
-#endif /* FREECIV_WEB */
-
-  if (
-#ifdef FREECIV_DEBUG
-      !CreateProcess(NULL, cmdline1, NULL, NULL, TRUE,
-                     DETACHED_PROCESS | NORMAL_PRIORITY_CLASS, NULL, NULL,
-                     &si, &pi)
-#ifndef FREECIV_WEB
-      && !CreateProcess(NULL, cmdline2, NULL, NULL, TRUE,
-                        DETACHED_PROCESS | NORMAL_PRIORITY_CLASS, NULL, NULL,
-                        &si, &pi)
-#endif /* FREECIV_WEB */
-      &&
-#endif /* FREECIV_DEBUG */
-      !CreateProcess(NULL, cmdline3, NULL, NULL, TRUE,
-                     DETACHED_PROCESS | NORMAL_PRIORITY_CLASS, NULL, NULL,
-                     &si, &pi)
-      && !CreateProcess(NULL, cmdline4, NULL, NULL, TRUE,
-                        DETACHED_PROCESS | NORMAL_PRIORITY_CLASS, NULL, NULL,
-                        &si, &pi)) {
-    log_error("Failed to start server process.");
-#ifdef FREECIV_DEBUG
-    log_verbose("Tried with commandline: '%s'", cmdline1);
-    log_verbose("Tried with commandline: '%s'", cmdline2);
-#endif /* FREECIV_DEBUG */
-    log_verbose("Tried with commandline: '%s'", cmdline3);
-    log_verbose("Tried with commandline: '%s'", cmdline4);
-    output_window_append(ftc_client, _("Couldn't start the server."));
-    output_window_append(ftc_client,
-                         _("You'll have to start one manually. Sorry..."));
-    return FALSE;
+  for (auto trueServer : qAsConst(program)) {
+    serverProcess::i()->start(trueServer, arguments);
+    trueFcser = trueServer;
+    if (serverProcess::i()->waitForStarted(3000) == true) {
```

Comment:
(Question not comment) Does this block the event loop or is it in another thread?

Reply:
it blocks